src/main/scala/beam/agentsim/agents/vehicles/VehicleEnergy.scala

```diff
@@ -0,0 +1,209 @@
+    overallFallBack: IndexedSeq[BeamVehicle.FuelConsumptionData] => Double,
+    singleFallBack: BeamVehicle.FuelConsumptionData => Double
+  ): Double = {
+    val consumptionsInJoules: IndexedSeq[Double] = fuelConsumptionDatas
+      .flatMap(fuelConsumptionData => {
+        for {
+          rateInJoulesPerMeter <- getRateUsing(fuelConsumptionData, singleFallBack)
+          distance             <- fuelConsumptionData.linkLength
+        } yield rateInJoulesPerMeter * distance
+      })
+    if (consumptionsInJoules.nonEmpty) consumptionsInJoules.sum else overallFallBack(fuelConsumptionDatas)
+  }
+
+  private def getRateUsing(fuelConsumptionData: BeamVehicle.FuelConsumptionData,
+                           fallBack: BeamVehicle.FuelConsumptionData => Double): Option[Double] = {
+    val BeamVehicle.FuelConsumptionData(linkId, vehicleType, numberOfLanesOption, _, _, _, speedInMilesPerHourOption, _, _, _) =
+      fuelConsumptionData
+    val numberOfLanes: Int = numberOfLanesOption.getOrElse(0)
+    val speedInMilesPerHour: Double = speedInMilesPerHourOption.getOrElse(0)
+    val gradePercent: Double = linkIdToGradePercentMap.getOrElse(linkId, 0)
+    consumptionRateFilterStore.getPrimaryConsumptionRateFilterFor(vehicleType).flatMap(
+      consumptionRateFilterFuture => getRateUsing(consumptionRateFilterFuture,
+        numberOfLanes, speedInMilesPerHour, gradePercent)
+    ).orElse(
+      consumptionRateFilterStore.getSecondaryConsumptionRateFilterFor(vehicleType).flatMap(
+        consumptionRateFilterFuture => getRateUsing(consumptionRateFilterFuture,
+          numberOfLanes, speedInMilesPerHour, gradePercent)
+      )
+    ).orElse(Some(fallBack(fuelConsumptionData)))
+  }
+
+  private def getRateUsing(consumptionRateFilterFuture: Future[ConsumptionRateFilter],
+                            numberOfLanes: Int, speedInMilesPerHour: Double, gradePercent: Double): Option[Double] = {
+    //1.)Future performance improvement could be to better index the bins so could fuzzily jump straight to it
+    //instead of having to iterate
+    //2.)Could keep the future in the calling method if you use
+    //Future.sequence and Option.option2Iterable followed by a flatMap(_.headOption),
+    //but that gets complicated and these SHOULD already be loaded by the time they are needed.
+    //If that changes then go ahead and map through the collections
+    import scala.concurrent.duration._
+    val consumptionRateFilter = Await.result(consumptionRateFilterFuture, 1.minute)
+    val filteredRates = consumptionRateFilter
+      .collect {
+        case (speedInMilesPerHourBin, restOfFilter) if speedInMilesPerHourBin.hasDouble(speedInMilesPerHour) =>
+          restOfFilter
+      }
+      .flatten
+      .collect { case (gradePercentBin, restOfFilter) if gradePercentBin.hasDouble(gradePercent) => restOfFilter }
+      .flatten
+      .collect { case (numberOfLanesBin, rate) if numberOfLanesBin.has(numberOfLanes) => rate }
```

Comment:
I'm worried about all of the processing required to just get a single number out of a table... this has to happen a LOT of times, for every link that every vehicle traverses. So 100's of millions of times in a single run. Performance is going to be critical, is this the fastest way to access the data?

Reply:
E.g. how would this approach with collect/flatten/etc compare to just putting all of the keys into a single n-tuple and make a single map lookup.

Map[(Int,Int,...),Double]

Reply:
Discussed online and resolved. There was a mental mismap on my part where I had expected possible overlaps of ranges which would resolve by the overall lookup. Note that this code will no longer be possible to catch bad data:

```
if (ratesSize > 1)
      log.warn(
        s"More than one ($ratesSize) rate was found using: Number of Lanes = $numberOfLanes; Speed in MPH = $speedInMilesPerHour; Grade Percent = $gradePercent." +
          "The first will be used, but the data should be reviewed for range overlap."
      )
```

Since the new return immediately as a hit via `find` will just ignore them. The same outcome, but no warning log. I'm sure that's fine - more of an FYI